x/wasm/internal/keeper/keeper.go

```diff
@@ -72,14 +74,34 @@ func NewKeeper(cdc *codec.Codec, storeKey sdk.StoreKey, accountKeeper auth.Accou
 		bankKeeper:    bankKeeper,
 		messenger:     messenger,
 		queryGasLimit: wasmConfig.SmartQueryGasLimit,
-		authZSchema:   DefaultAuthorizationSchema{},
+		authZPolicy:   DefaultAuthorizationPolicy{},
+		paramSpace:    ps,
 	}
 	keeper.queryPlugins = DefaultQueryPlugins(bankKeeper, stakingKeeper, keeper).Merge(customPlugins)
 	return keeper
 }
 
+func (k Keeper) getUploadAccessConfig(ctx sdk.Context) types.AccessConfig {
+	var a types.AccessConfig
+	k.paramSpace.Get(ctx, types.ParamStoreKeyUploadAccess, &a)
+	return a
+}
+
+func (k Keeper) getInstantiateAccessConfig(ctx sdk.Context) types.AccessType {
+	var a types.AccessType
+	k.paramSpace.Get(ctx, types.ParamStoreKeyUploadAccess, &a)
+	return a
+}
+
 // Create uploads and compiles a WASM contract, returning a short identifier for the contract
 func (k Keeper) Create(ctx sdk.Context, creator sdk.AccAddress, wasmCode []byte, source string, builder string) (codeID uint64, err error) {
+	return k.create(ctx, creator, wasmCode, source, builder, k.authZPolicy)
+}
+
+func (k Keeper) create(ctx sdk.Context, creator sdk.AccAddress, wasmCode []byte, source string, builder string, authZ AuthorizationPolicy) (codeID uint64, err error) {
+	if !authZ.CanCreateCode(k.getUploadAccessConfig(ctx), creator) {
```

Comment:
Very nice here.

I would change the message from "no permission" to "cannot create" or such, as ErrUnauthorized default message is the same as "no permission" basically